_posts/2024-07-01-zio-fullstack-webapp.md

```diff
@@ -0,0 +1,844 @@
+
+
+
+```scala
+trait UserRepository: // (1)
+  def create(user: User): Task[User]
+  def getById(id: Long): Task[Option[User]]
+ 
+                                                 // (2)
+class UserRepositoryLive private (quill: Quill.Postgres[SnakeCase]) extends UserRepository {
+
+  import quill.*
+
+  inline given SchemaMeta[User] = schemaMeta[User]("users")  // (3)
+  inline given InsertMeta[User] = insertMeta[User](_.id)
+  inline given UpdateMeta[User] = updateMeta[User](_.id, _.creationDate)
+
+  override def create(user: User): Task[User] =
+    run(query[User].insertValue(lift(user)).returning(r => r)) // (4a)
+  override def getById(id: Long): Task[Option[User]] =         // (4b)
+    run(query[User].filter(_.id == lift(Option(id)))).map(_.headOption)
+}
+
+object UserRepositoryLive: // (5)                                        (6)
+  def layer: RLayer[Postgres[SnakeCase], UserRepositoryLive] = ZLayer.derive[UserRepositoryLive]
+
+```
+
+* **(1)** Repository trait
+* **(2)** Repository implementation, with Quill Postgres injection
+* **(3)** Quill Meta, to map the User case class to the users table
+* **(4)** Quill queries
+* **(5)** Repository layer
+* **(6)** One of the many ZIO macro that will generate the layer for you, from the constructor of the class (2).
+
+
+💡 Note how `ZLayer.derive[UserRepositoryLive]` is used to generate the layer from the constructor of the class, and how this layer is then depends on the `Postgres[SnakeCase]`.
+
+
+Quill is a compile-time query language for Scala and SQL. It allows to write queries in Scala and compile them to SQL.
+
+
+Quill compiles **(4a)** and **(4b)** to SQL like this:
+
+
+```sql
+INSERT INTO users (name,email,age,creation_date) VALUES (?, ?, ?, ?) RETURNING id, name, email, age, creation_date;
+SELECT x4.id, x4.name, x4.email, x4.age, x4.creation_date AS creationDate FROM users x4 WHERE x4.id IS NULL AND ? IS NULL OR x4.id = ?;
+```
+
+
+### 5.1.3. Flyway Service
+
+In the same way, Flyway service is a layer that will handle the database migrations.
+
+```scala
+object FlywayServiceLive {
+  def live: RLayer[FlywayConfig, FlywayService] = ZLayer( // (1)
+    for {
+      config <- ZIO.service[FlywayConfig]
+      flyway <- ZIO.attempt(Flyway.configure()
+                   .dataSource(config.url, config.user, config.password).load())
+    } yield new FlywayServiceLive(flyway)
+  )
+
+  val configuredLayer: TaskLayer[FlywayService] = Configs.makeConfigLayer[FlywayConfig]("db.dataSource") 
+                                                             >>> live  // (2)
+}
+```
+  
+* **(1)** is a layer that will provide a **FlywayService**, it depends on the availability of **FlywayConfig** service.
+* **(2)** Inject a **FlywayConfig** in a the previous layer and the remove the dependency on the config.
+
+### 5.1.4. Person Service
+
+Person service is a layer that will handle the business logic of the application.
```

Comment:
You can add a simple description of what this service layer does. You can say that it's common for the service to delegate a lot of functionality to the repository layer, but the business logic can be arbitrarily complex.